### PR TITLE
🔒. Hashed ElGamal Encryption

### LINKED ISSUE #260: ✨ Add a HashedElGamalCiphertext object for encryption (and decryption - currently for testing) for overvotes and writeins
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Suggestion

 An implementation of the Auxiliary Encryption as described in the ElectionGuard specification is needed to encrypt (and decrypt - currently just for testing) overvotes and writeins. 

### Possible Implementation

Create a HashedElGamalCiphertext object for encryption (and decryption - currently for testing) for overvotes and writeins. The encryption will use the Auxiliary Encryption as specified in the ElectionGuard specification.

The current padding scheme agreed on is as follows:
Two ways to encrypt.

1) If the plaintext is fixed length and a multiple of the block length (32 bytes) then the caller may indicate this and no padding is done. The code throws an exception in this case if the plaintext is not a multiple of the block length.

2) If the plaintext length is variable then a 4 byte length field is concatenated to the front of the plaintext in big endian form. This data is then padded out with 0x00 bytes to the block length. If the 4 byte length field plus the plaintext length falls on a block length boundary (for example if the plaintext is 28 bytes then the with the length concatenated then it makes it 32 bytes) then no 0x00 padding is applied. On decrypt the length is checked to make sure it is not longer than the ciphertext length minus the 4 byte length. If it is then an exception is thrown. The padding length is calculated and if any padding bytes are non 0x00 then an exception is thrown.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jeffspel-crypto**
[//]: # (🚨 Please review the CONTRIBUTING.md in this repository. 💔Thank you!)

### Issue
*Link your PR to an issue*

Fixes #260

### Description
This change adds an object which is HashedElGamalCiphertext that supports encrypt/decrypt as specified for Auxiliary Encryption in the ElectionGuard specification. This object will eventually be used to encrypt overvotes and writeins at the encryptConstest level. There is not yet support for precomputed values but that will be done once the work to encrypt overvotes and writeins is complete.

### Testing
Success test cases have been added for encrypt/decrypt of data that is not padded as well as data that requires padding. Failure cases have been added for different nonces on encrypt/decrypt as well as tampered ciphertext. A benchmark test has been added for the encrypt method.

### CODE REVIEW

File: src/electionguard/elgamal.cpp [hunk 1]
```diff
@@ -258,4 +264,388 @@ namespace electionguard
...
+
+        Impl(unique_ptr<ElementModP> pad, vector<uint8_t> ciphertext, vector<uint8_t> mac)
+            : pad(move(pad)), ciphertext(ciphertext), mac(mac)
+        {
+        }
+
+        [[nodiscard]] unique_ptr<HashedElGamalCiphertext::Impl> clone() const
+        {
+            auto _pad = make_unique<ElementModP>(*pad);
+            return make_unique<HashedElGamalCiphertext::Impl>(move(_pad), ciphertext, mac);
```
Comment (SteveMaier-IRT) [inline src/electionguard/elgamal.cpp:283]: Can we just use Clone here instead of making a new smart pointer and then moving the data?

Comment (jeffspel-crypto) [reply src/electionguard/elgamal.cpp:283]: Sure, we can do that. I was following the convention used in the ElGamalCiphertext::Impl above, but either way works for me.

Comment (SteveMaier-IRT) [reply src/electionguard/elgamal.cpp:283]: I am reviewing and looking into what was done before to make sure that we are sure of the patterns used.

**Comment by keithrfung** [comment]
@SteveMaier-IRT Be sure to keep the renamed name on any squash and merge.

File: src/electionguard/elgamal.cpp [hunk 2]
```diff
@@ -258,4 +264,388 @@ namespace electionguard
         return make_unique<ElGamalCiphertext>(move(resultPad), move(resultData));
     }
 
+
+#pragma region HashedElGamalCiphertext
+    
+    struct HashedElGamalCiphertext::Impl {
+        unique_ptr<ElementModP> pad;
+        vector<uint8_t> ciphertext;
```
Comment (keithrfung) [inline src/electionguard/elgamal.cpp:272]: We label this as `data` in Python

Comment (jeffspel-crypto) [reply src/electionguard/elgamal.cpp:272]: Sure, I will make that change.

**Pull request closed without merging**